# Regresia liniară de la zero

Regresia liniară prezice o valoare numerică folosind o dreaptă. Vom estima timpul de rezolvare al unei probleme în funcție de numărul de exerciții similare lucrate.

## 1. Modelul

Pentru o intrare $x$, modelul produce predicția $\hat{y} = wx + b$. Parametrul $w$ controlează panta, iar $b$ este valoarea estimată când $x = 0$.

Un model este util doar dacă funcționează pe exemple pe care nu le-a văzut în timpul antrenării.

In [1]:
exercitii = [1, 2, 3, 4, 5]
minute = [55, 50, 43, 35, 27]

w, b = -8, 65
predictii = [w * x + b for x in exercitii]
print("Predicții:", predictii)

Predicții: [57, 49, 41, 33, 25]


## 2. Măsurăm eroarea

Eroarea medie absolută — MAE — este media distanțelor dintre predicții și valorile reale:

$$\operatorname{MAE} = \frac{1}{n} \sum_{i=1}^{n} |y_i - \hat{y}_i|$$

Cu cât valoarea este mai mică, cu atât predicțiile sunt mai apropiate de date.

In [2]:
def mae(valori_reale, valori_estimate):
    erori = [abs(y - y_hat) for y, y_hat in zip(valori_reale, valori_estimate)]
    return sum(erori) / len(erori)

print(f"MAE: {mae(minute, predictii):.1f} minute")

MAE: 1.8 minute


## 3. Experimentează

Schimbă valorile lui `w` și `b`. Poți obține o eroare mai mică de 1.8? Notează cea mai bună combinație găsită. Acesta este, în esență, procesul de **antrenare**: căutarea parametrilor care minimizează o funcție de eroare.

> Datele sunt intenționat puține pentru a putea inspecta fiecare calcul. Într-un proiect real vom separa datele în seturi de antrenare și testare.